[Prototype] Filter MLv1 <> MLv2 converstion

kulyk opened the pull request.

bshepherdson committed:
[MLv2] Add functions to support viz settings in FE

Viz settings contain legacy MLv1 field refs for the selected columns of
a query. We need to round-trip those settings into MLv2 `:fields` lists
and back to legacy refs.

This adds:
- `visible-columns` to get the full set of columns in scope on a query
  - this includes `:selected?` information for the columns
- `legacy-field-ref` to convert an MLv2 field ref to legacy MLv1
- `find-column-index-from-legacy-ref` to match a legacy ref against a
  list of MLv2 columns

bshepherdson committed:
review comments

bshepherdson committed:
refactoring to find many at once

ranquild committed:
Move changing query columns to MLv2 (#32563)

ranquild committed:
Fix issue with orphan table settings

ranquild committed:
Add a test for multi-stage queries

bshepherdson committed:
all call sites refactored

bshepherdson committed:
fix all BE tests

bshepherdson committed:
fix some BE-side bugs - aggregation refs work now

I think custom expressions are still busted

ranquild committed:
Use new find by index method

ranquild committed:
Simplify working with broken field refs

ranquild committed:
Remove syncTableColumnsToQuery

bshepherdson committed:
fix the return value of findColumnIndexesFromLegacyRefs

bshepherdson committed:
fix return value for findColumnIndexesFromLegacyRefs

ranquild committed:
Fix index issues

ranquild committed:
Unskip tests

bshepherdson committed:
add/remove-field should work normally for expressions

Previously custom expressions were a special case and not included in
`:fields` lists; this was an error and now they work like any other
column.

bshepherdson committed:
findColumnIndexesForLegacyRefs: more robust conversion

ranquild committed:
FE fixes with aggregations & breakouts

ranquild committed:
Fix e2e test

bshepherdson committed:
Added implicitly joinable columns with `:source-card` queries

Most of table-column-settings.cy.spec.js is now passing, remaining issue
is I think because custom expressions are not included from
`:source-card` queries but should be.

This has failing BE unit tests but I'm about to leave for vacation.

https://metaboat.slack.com/archives/C04CYTEL9N2/p1691676822350829
Slack thread for much more context on the state of this PR.

metamben committed:
Fix BE tests

metamben committed:
Fix test in table-column-settings.cy.spec.js

metamben committed:
Merge remote-tracking branch 'origin/master' into mblib-column-settings-round-trip

metamben committed:
Merge remote-tracking branch 'origin/master' into mblib-column-settings-round-trip

metamben committed:
Fix kondo reported issues

metamben committed:
Fix typing error

metamben committed:
Merge branch 'master' into mblib-column-settings-round-trip

ranquild committed:
Fix passing role

ranquild committed:
Fix MLv1 vs MLv2 dimension comparison

ranquild committed:
Fix the test

ranquild committed:
Fix the e2e test

ranquild committed:
Fix settings e2e test

ranquild committed:
Remove datasetColumn check

ranquild committed:
Remove datasetColumn check

ranquild committed:
Fix unit test

kulyk committed:
Fix MLv2 types

kulyk committed:
Expose structured query's `getMLv2Query`

kulyk committed:
Fix filter `arguments` method return type

kulyk committed:
Prototype filter conversion utils

iethree commented on frontend/src/metabase-lib/compat/compat.ts:
I like the use of a compat/ directory 👍

metamben committed:
Merge branch 'master' into mblib-column-settings-round-trip

iethree reviewed (commented):
I think this is essentially a better version of what I tried to do [here](https://github.com/metabase/metabase/commit/356ddca7218b9cd8a5b6e2344c3f424e22e5402d), and decided was [fundamentally a bad idea](https://metaboat.slack.com/archives/C04CYTEL9N2/p1691446188288369?thread_ts=1690925869.951219&cid=C04CYTEL9N2).

But I still have a few misgivings:
1. I think `externalOp` is a poor API for use outside of metabase-lib internals because it doesn't tell you what it does at all.
2. I'm not confident that we're actually correctly converting these things: especially when it comes to columns/fields/dimensions 🥴 
3. By doing this in JS-land, we're essentially building MLv1.5 on top of MLv1, and taking a performance hit for it, rather than dumping MBQL into CLJC and letting the QP code handle it. I think this exposes a challenge in building the frontend with two different teams and codebases. It's a little hard to determine where a bit of logic _should_ live, and it often easier to jump through hoops and just get it done sub-optimally, than request a new API from another team. I think this is one of the patterns we were trying to avoid with this project.

All that said, I like the use of `compat` functions, and we've got to move forward with this project sooner rather than later. If we can nail down a good external API for the `compat` functions (🙏  please no externalOp 🙏 ) , we can muck around with their internals or move them to CLJC if we find that they're inadequate.

kulyk commented on frontend/src/metabase-lib/compat/utils.ts:
```diff
@@ -0,0 +1,75 @@
+  const columnName = displayInfo(query, stageIndex, column).longDisplayName;
+  return filterableColumns(query, stageIndex).find(col => {
+    const info = displayInfo(query, stageIndex, col);
+    return info.longDisplayName === columnName;
```
We won't actually need this function (as most of the stuff in `compat/utils`) if `externalOp` returns a column as `ColumnWithOperators` instead of `ColumnMetadata`

We shouldn't do these checks, but comparing `longDisplayName` is the most robust way to compare columns so far (fortunately we didn't need that before, haha). Comparing field IDs won't work because not every column has an ID (e.g. aggregated or custom columns). OTOH, `longDisplayName` has a table name for foreign columns, so the risk of having two identical names is low. We should avoid using it for real though

kulyk commented on frontend/src/metabase-lib/compat/utils.ts:
```diff
@@ -0,0 +1,75 @@
+}
+
+// This won't be needed once externalOp returns `operator` as an opaque object
+export function findOperator(
+  query: Query,
+  column: ColumnMetadata | ColumnWithOperators,
+  shortName: string,
+) {
+  // This won't be needed once externalOp columns are of `ColumnWithOperators` type
+  const columnWithOperators = findFilterableColumn(query, column);
+  if (!columnWithOperators) {
+    return null;
+  }
+
+  const operators = filterableColumnOperators(columnWithOperators);
+  const operator = operators.find(operator => {
+    const info = displayInfo(query, stageIndex, operator);
+    return info.shortName === shortName;
+  });
+  return operator || null;
+}
+
+export function fixExternalOp(query: Query, filter: FilterClause) {
+  const {
+    operator: operatorName,
+    options,
+    args: [column, ...args],
+  } = externalOp(filter);
+
+  const operator = findOperator(query, column, operatorName);
+
+  // This won't be needed once externalOp columns are of `ColumnWithOperators` type
+  const columnWithOperators = findFilterableColumn(query, column) || null;
+
+  return {
+    operator,
+    options,
+    args: [columnWithOperators, ...args],
+  };
+}
```
I was trying to communicate that with `"This won't be needed once..."` comments:

- `findOperator` will become redundant if `externalOp` returns an opaque operator
- `fixExternalOp` will become redundant if`externalOp` returns an opaque operator and returns `ColumnWithOperators` instead of `ColumnMetadata`

kulyk commented on frontend/src/metabase-lib/compat/compat.ts:
```diff
@@ -0,0 +1,115 @@
+      return {
+        query,
+        filter,
+        externalOp: fixExternalOp(query, filter),
```
I don't have a strong preference, but there's a high chance I'm already used to `externalOp`. Definitely agree the name isn't clear without extra context

iethree replied:
IMHO, one of the biggest benefits of MLv2 should be to make it easier to use, especially for new devs with little or no context.

kulyk commented:
Thanks @iethree 🙏 

> I think `externalOp` is a poor API for use outside of metabase-lib internals because it doesn't tell you what it does at all

I'm not sure if you have concerns about the name (could be something like `filterParts` or whatever), and/or the overall concept of grouping filter parts into a single object?

> I'm not confident that we're actually correctly converting these things: especially when it comes to columns/fields/dimensions

Could you please clarify what exactly feels fragile? MLv1 <> MLv2 columns conversion here is done on the CLJC side via `legacyFieldRef` and `findColumnIndexesFromLegacyRefs` methods

> By doing this in JS-land, we're essentially building MLv1.5 on top of MLv1, and taking a performance hit for it, rather than dumping MBQL into CLJC and letting the QP code handle it. I think this exposes a challenge in building the frontend with two different teams and codebases

1. Performace. Please take a look at my comments above, I believe what might have looked like a performance drop is actually a temporary prototype code (most things under `compat/utils`)
2. Moving more stuff to the BE part. To me, it seems like this problem has two parts: applied and draft filters. Draft filters are the hard part because MLv2 is designed to always produce valid queries and IMO it's good if we could keep it away from partially complete MLv1 data. @metamben also confirmed it's going to be difficult to handle. OTOH, for applied filters things are way more straightforward because we can just get a MLv2 object by filter index and it'd be solid enough. So we have a draft filter problem that has to be partially handled on the FE anyway and a straightforward applied filter scenario. IMO we could just handle both cases in one helper without significant risk.

metamben committed:
Fix bad merge

metamben committed:
Merge remote-tracking branch 'origin/master' into mblib-column-settings-round-trip

metamben committed:
Merge remote-tracking branch 'origin/master' into mblib-column-settings-round-trip

metamben committed:
Revert tests restricting cards' visible-columns to returned-columns

ranquild committed:
Add repro for the 1st issue

ranquild committed:
Add repro for the 2nd issue

ranquild committed:
Add repro for the 3rd issue

metamben committed:
Fix indentation

metamben committed:
Fix matching MLv1 and MLv2 columns and references

metamben committed:
Merge remote-tracking branch 'origin/master' into mblib-column-settings-round-trip

metamben committed:
Allow removal of implicitly joinables w/o explicit fields clause

As demonstrated by metabase#32373 in table-column-settings.cy.spec.js, fields coming from source questions can be
implicitly joinable according to :lib/source and still appear without an explicit :fields clause.
